run/generic_runner.go

```diff
@@ -0,0 +1,219 @@
+	runnerConfig any,
+) (deferFunc func() error, err error) {
+	deferFunc = func() error {
+		return nil
+	}
+	if cpuProfiler, ok := runnerConfig.(CPUProfiler); ok &&
+		cpuProfiler.CPUProfilePath() != "" {
+		// CPU profiler.
+		f, err := os.Create(cpuProfiler.CPUProfilePath())
+		if err != nil {
+			return deferFunc, err
+		}
+		deferFunc = func() error {
+			return f.Close()
+		}
+
+		if err := pprof.StartCPUProfile(f); err != nil {
+			return deferFunc, err
+		}
+		defer pprof.StopCPUProfile()
+	}
+	return deferFunc, nil
+}
+
+func (r *genericRunner[RunnerConfig, Input, Option, Solution],
+) handleMemoryProfile(runnerConfig any,
+) (deferFunc func() error, err error) {
+	deferFunc = func() error {
+		return nil
+	}
+	// Memory profile.
+	if memoryProfiler, ok := runnerConfig.(MemoryProfiler); ok &&
+		memoryProfiler.MemoryProfilePath() != "" {
+		f, err := os.Create(memoryProfiler.MemoryProfilePath())
+		if err != nil {
+			return deferFunc, err
+		}
+		deferFunc = func() error {
+			return f.Close()
+		}
+
+		// Clean up unused objects from the heap before profiling. But do not
+		// garbage collect the runner, so we can see in-use memory.
+		runtime.GC()
+		runtime.KeepAlive(r)
+
+		if err := pprof.WriteHeapProfile(f); err != nil {
+			return deferFunc, err
+		}
+	}
+	return deferFunc, nil
+}
+
+func (r *genericRunner[RunnerConfig, Input, Option, Solution]) Run(
+	context context.Context,
+) (retErr error) {
+	// handle CPU profile
+	deferFuncCPU, retErr := r.handleCPUProfile(r.runnerConfig)
+	if retErr != nil {
+		return retErr
+	}
+	defer func() {
+		err := deferFuncCPU()
+		// the first error is more important
+		if retErr == nil {
+			retErr = err
+		}
+	}()
+	// get IO
+	ioData, retErr := r.IOProducer(context, r.runnerConfig)
+	if retErr != nil {
+		return retErr
+	}
+
+	// decode input
+	decodedInput, retErr := r.InputDecoder(context, ioData.Input())
+	if retErr != nil {
+		return retErr
+	}
+
+	// use options configured in runner via flags and environment variables
+	decodedOption := r.flagParsedOption
+	// decode option if provided
+	tempOption, err := r.OptionDecoder(context, ioData.Option())
+	if err != nil {
+		return err
+	}
+	var defaultOption Option
+	// if option is not default, use it
+	if !reflect.DeepEqual(tempOption, defaultOption) {
+		decodedOption = tempOption
+	}
+
+	// run algorithm
+	solutions := make(chan Solution)
+	errs := make(chan error, 1)
+	go func() {
+		defer close(solutions)
+		defer close(errs)
+		retErr = r.Algorithm(context, decodedInput, decodedOption, solutions)
+		if retErr != nil {
+			errs <- retErr
+			return
+		}
+	}()
+
+	// encode solutions
+	retErr = r.Encoder.Encode(
+		context, solutions, ioData.Writer(), r.runnerConfig, decodedOption,
+	)
+	if retErr != nil {
+		return retErr
+	}
+
+	// handle memory profile
+	deferFuncMemory, retErr := r.handleMemoryProfile(r.runnerConfig)
+	if retErr != nil {
+		return retErr
+	}
+	defer func() {
```

Comment:
We are already at the end of the function. There is no need for a `defer` call I think

Reply:
The alternative code should be

```go
	retErr = <-errs
	err = deferFuncMemory()
	if retErr == nil {
		retErr = err
	}

	return retErr
```

but I am uncertain whether that makes the intention more clear.

Reply:
Or move it up to the CPU profile?

Reply:
We can't do that because we want to profile after the run to see memory consumption at that time.